### PR TITLE
[fix] Try something to be independent from github

### PR DESCRIPTION
**Pull request by frju365**
## Problem
This PR refers to https://github.com/YunoHost/issues/issues/1117. We can only install package from github.

## Solution
Add a condition which will test a connection to different API, and that trying to send GET requests to upstream. This PR implements GET command for Gitlab CE and Gitlab.com based-repo.
 
## Status
Work in progress. Don't Merge.

### CODE REVIEW

**Comment by alexAubin** [comment]
(Bump?)

**Comment by frju365** [comment]
mh... yes. My condition is missing or false. I'm working on that. I hope pusing my modifications tomorrow or this weekend.

File: src/views/app/app_list_install.ms [hunk 1]
```diff
@@ -52,7 +52,7 @@
             <div class="form-group has-feedback">
                 <label for="url" class="col-sm-12">{{t 'url'}}</label>
                 <div class="col-sm-12">
-                    <input type="url" id="url" name="url" class="form-control" value="" placeholder="https://github.com/USER/REPOSITORY" required pattern="^https://github.com/[a-zA-Z0-9-_.]+/[a-zA-Z0-9-_.]+[/]?$">
+                    <input type="url" id="url" name="url" class="form-control" value="" placeholder="https://GITHOST.TLD/USER/REPOSITORY" required pattern="^https?://.+/[a-zA-Z0-9-_.]+/[a-zA-Z0-9-_.]+[/]?$">
```
Comment (opi) [inline src/views/app/app_list_install.ms]: Maybe remove the `pattern`, as one could host a gitlab on "https://domain.com/my-gitlab" and then a repo url will be "https://domain.com/my-gitlab/namespace/repo" ; 
Or, this pattern could be nice to ensure that repo is served with httpS support.

**Comment by frju365** [comment]
I stoped the work on this PR for instance as I have much work (and took all my time). I will continue during the hollydays

**Comment by isAAAc** [comment]
why not hosting our own distributed gitea repos ? gitea is packaged for ynh

**Comment by alexAubin** [comment]
Gitea is lacking quite a lot of features that would be useful (e.g in terms of PR reviews at least I think). Last time we discussed this, we concluded that Gitlab would be the best option to get rid of Github. Also having our own's forge increase maintenance costs and contribution cost as one needs to create an account on the forge for the slightest interaction. So we concluded we'd be using Framagit probably.

But setting up / choosing an alternative forge to Github really is the easy part of the problem. The biggest issue is taking the time to migrate everything we have to other forge. While discussing it, we concluded that we have at least 10 piece of softwares tightly couples to Github so :/ ...

**Comment by frju365** [comment]
Well migration is like this from my point of view : 
1st : we fix Yunohost-admin to accept repos from Gitlab/Gitea/Notabug/framagit... Well other Github-like software.
2d We tell users and packager that  OFFICIALLY accept other and we can begin to move official app and tell packager they can easily move them
3d We fix Yunohost core (the most difficult)

**Comment by frju365** [comment]
Well, normally it finished, as quite all GitHost have the same RAW path for sources preview. Tested but installed was not possible because of Cross-Domain policy in my local test

**Comment by frju365** [comment]
Any update on this PR ? Normally I finished it. If someone want to test it... :)

**Comment by alexAubin** [comment]
Sounds alright, to be discussed and tested after we're past the 3.7

**Comment by alexAubin** [comment]
Yolomergin' to test and validate

**Comment by alexAubin** [comment]
Annnnd "how bad can it be", well then I couldnt get any install to work, even the regular ones from Github. Tested also with a repo hosted on framagit, and with https://code.antopie.org/miraty/qr_ynh. Ajax just returns an error for some reason. Reverting for now. Feel free to resubmit when it'll actually be working ...

**Comment by frju365** [comment]
Do you remember what was this error ? Is it the error I mentioned in my last comment ?

**Comment by alexAubin** [comment]
I had some initial issue with cross-domain because of some browser add-on, yes. But then I disabled it and still got some error ... There was no error message in particular unfortunately. I just did some dirty debugging, identified that the code ran into the `.fail()` part. The `xhr` object just contained something like `statusCode: error` ... despite the fact that in the Network tab of Firefox, the request appeared succesful (code 200 + apparently relevant content)

For https://code.antopie.org/miraty/qr_ynh, I identified that the server throws a redirect (302) which apparently ajax doesn't like and therefore fails somehow...

**Comment by NathanRussellUK** [comment]
Hello,

This sounds great - just wondering where the thinking's at for the moment, as it'd be great to use the UI to install custom apps!

At the moment, I'm using the CLI (i.e. `sudo yunohost app install [url to git repo]`) and this is working for custom apps on my self-hosted gitlab. This is fine but a bit time-consuming doing it myself! Could the web interface call this command in some way?

Thanks for everything!

Nathan

**Comment by alexAubin** [comment]
Yes, we need to rethink the way this works. Basically no ajax request should be done inside the browser (because of a) CSP and b) because we can't assume the raw json we'll fetch is alright). Imho we need an API route like `GET /apps/download/<raw_url>` that would run some git clone in the backend and return the manifest in the json data to the client.

**Pull request merged**